### PR TITLE
UCT/MD: added md_dereg_v2 implementation

### PR DESCRIPTION
**Pull request by hoopoepg**
- added uct_md_mem_dereg_v2 implementation
- added support of memory invalidation into IB memory
  domain
- improved UCT_MD_MEM_DEREG_CHECK_PARAMS macro
- added gtest for new API

### CODE REVIEW

File: src/uct/base/uct_md.h [hunk 1]
```diff
@@ -33,10 +33,17 @@
...
-    if (ENABLE_PARAMS_CHECK && !(_invalidate_supported) && \
-        (UCT_MD_MEM_DEREG_FIELD_VALUE(_params, flags, FIELD_FLAGS, 0) & \
-                UCT_MD_MEM_DEREG_FLAG_INVALIDATE)) { \
-        return UCS_ERR_UNSUPPORTED; \
+    if (ENABLE_PARAMS_CHECK) { \
+        if (UCT_MD_MEM_DEREG_FIELD_VALUE(_params, flags, FIELD_FLAGS, 0) & \
+            UCT_MD_MEM_DEREG_FLAG_INVALIDATE) { \
+            if (!(_invalidate_supported)) { \
+                return UCS_ERR_UNSUPPORTED; \
+            } else if (!UCT_MD_MEM_DEREG_FIELD_VALUE(params, cb, \
```
Comment (yosefe) [inline src/uct/base/uct_md.h:41]: no need else

File: src/uct/ib/base/ib_md.c [hunk 2]
```diff
@@ -1037,9 +1043,17 @@ uct_ib_mem_rcache_dereg(uct_md_h uct_md,
...
-    UCT_MD_MEM_DEREG_CHECK_PARAMS(params, 0);
+    UCT_MD_MEM_DEREG_CHECK_PARAMS(params, 1);
 
     region = uct_ib_rcache_region_from_memh(params->memh);
+    if (UCT_MD_MEM_DEREG_FIELD_VALUE(params, flags, FIELD_FLAGS, 0) &
+        UCT_MD_MEM_DEREG_FLAG_INVALIDATE) {
+        ucs_rcache_region_invalidate(md->rcache, &region->super, params->cb,
+                                     UCT_MD_MEM_DEREG_FIELD_VALUE(params, arg,
+                                                                  FIELD_ARG,
+                                                                  NULL));
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:1054]: minor: use helper var for arg

File: test/gtest/uct/test_md.cc [hunk 3]
```diff
@@ -638,6 +646,117 @@ UCS_TEST_P(test_md, sockaddr_accessibility) {
...
+    uct_md_mem_dereg_params_t params;
+
+    params.field_mask = UCT_MD_MEM_DEREG_FIELD_FLAGS |
+                        UCT_MD_MEM_DEREG_FIELD_MEMH |
+                        UCT_MD_MEM_DEREG_FIELD_CALLBACK |
+                        UCT_MD_MEM_DEREG_FIELD_ARG;
+    params.cb         = dereg_cb;
+    params.arg        = this;
+
+    ptr               = malloc(size);
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:670]: wierd indent

File: src/uct/base/uct_md.h (hunk 1 above)
Comment (hoopoepg) [reply src/uct/base/uct_md.h]: ok, removed

File: src/uct/ib/base/ib_md.c (hunk 2 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: used helper var

File: test/gtest/uct/test_md.cc (hunk 3 above)
Comment (hoopoepg) [reply test/gtest/uct/test_md.cc:669]: aligned with previous ```=```

**Review by brminich — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

**Comment by hoopoepg** [comment]
@yosefe I fixed test on CUDA systems
ok to squash?

File: test/gtest/uct/test_md.cc [hunk 4]
```diff
@@ -700,7 +700,8 @@ UCS_TEST_SKIP_COND_P(test_md, invalidate, !check_caps(UCT_MD_FLAG_INVALIDATE))
     free(ptr);
 }
 
-UCS_TEST_SKIP_COND_P(test_md, dereg_bad_arg, !check_caps(UCT_MD_FLAG_REG))
+UCS_TEST_SKIP_COND_P(test_md, dereg_bad_arg,
+                     !check_reg_mem_type(UCS_MEMORY_TYPE_HOST))
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:704]: why is it needed?

Comment (hoopoepg) [reply test/gtest/uct/test_md.cc:704]: it failed to register memory on CUDA and test failed
will improve test later to cover all supported memory types in MD

Comment (yosefe) [reply test/gtest/uct/test_md.cc:704]: what is the error? could it be some real issue with invalidation impl ?

Comment (hoopoepg) [reply test/gtest/uct/test_md.cc:704]: it is failed on line 713 - status is not OK

Comment (yosefe) [reply test/gtest/uct/test_md.cc:704]: i see, 10x

**Review by yosefe — APPROVED** [review]

**Comment by hoopoepg** [comment]
will squash now

**Pull request merged**